samples/convert-upload-images-powervs/convert-upload-images-powervs

```diff
@@ -612,16 +628,164 @@ function copy_image_file {
 }
 
 function download_url() {
-  local url=$1
+  local url="$1"
+  local expected_sha256="$2"
   local image_name=${url##*/}
-  rm -rf $image_name
-  retry "curl -fsSL $url -o ./$image_name"
-  if [[ $? -eq 0 ]]; then
-    #IMAGE_PATH=$(realpath ./$image_name)
-    IMAGE_PATH=./$image_name
-    DOWNLOAD_IMAGE_NAME=$image_name
+  local retry_count=0
+  local download_success=false
+  
+  log "========================================="
+  log "Starting download: $(basename "$image_name")"
+  log "Source URL: $url"
+  log "========================================="
+  
+  # Validate URL before attempting download
+  validate_url "$url"
+  
+  # Remove any existing file
+  rm -rf "$image_name"
+  
+  # Retry loop
+  while [ $retry_count -lt $DOWNLOAD_MAX_RETRIES ]; do
+    if [ $retry_count -gt 0 ]; then
+      warn "Retry attempt $retry_count of $DOWNLOAD_MAX_RETRIES"
+      sleep $DOWNLOAD_RETRY_DELAY
+    else
+      log "Download attempt $((retry_count + 1)) of $DOWNLOAD_MAX_RETRIES"
+    fi
+    
+    # Download with timeout and progress
+    log "Downloading $(basename "$image_name")..."
+    if curl -fLSs --retry 2 --retry-delay 2 --connect-timeout 60 --max-time $DOWNLOAD_TIMEOUT "$url" -o "./$image_name" 2>&1; then
+      
+      # Verify file exists and has content
+      if [ ! -f "./$image_name" ] || [ ! -s "./$image_name" ]; then
+        error "Downloaded file is missing or empty"
+        retry_count=$((retry_count + 1))
```

Comment:
Once the file is downloaded, contents can rarely change over re-downloading. it's better to move this check outside the loop.